### PR TITLE
fixes #5098, properly detect has one relations

### PR DESCRIPTION
**Pull request by nineinchnick**
Fixes yiisoft/yii2#5098. Properly compare primary and all unique keys to foreign key columns.
Added unit tests.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
+nbproject
+
+# zend studio for eclipse project files
+.buildpath
+.project
+.settings
+
+# windows thumbnail cache
+Thumbs.db
+
```
Comment (cebe) [inline .gitignore]: this is not needed, you should have a gitignore global file for such things.
see https://gist.github.com/cebe/1166792#file-gitignore_global for an example.

File: generators/model/Generator.php [hunk 2]
```diff
@@ -323,15 +323,17 @@ protected function generateRelations()
...
-                    foreach ($fks as $key) {
-                        if (!in_array($key, $table->primaryKey, true)) {
-                            $hasMany = true;
-                            break;
-                        }
+                $uniqueKeys = [$table->primaryKey];
+                try {
+                    $uniqueKeys = array_merge($uniqueKeys, $db->getSchema()->findUniqueIndexes($table));
+                } catch (NotSupportedException $e) {
+                    // pass
```
Comment (samdark) [inline generators/model/Generator.php]: You mean ignore?

**Comment by samdark** [comment]
Looks good to me. Need changelog.

**Comment by nineinchnick** [comment]
@samdark this one's ready for merging.

**Comment by samdark** [comment]
Merged https://github.com/yiisoft/yii2-gii/commit/00719bfd11760c8f442b87ff24b858339e753bfe. Thanks!

**Comment by WadeShuler** [comment]
I think Gii is still generating the wrong relation.. Wanted to mention it here before opening a new issue, just to be sure it isn't something I should have done differently.

I used the include init migration to create the user table. I then created a new migration to create a user_profile table.

    $this->createTable('{{%user_profile}}', [
            'id' => $this->primaryKey(),
            'user_id' => $this->integer()->notNull(),
            'nicename' => $this->string(30)->notNull(),
            'use_gravatar' => $this->boolean()->defaultValue(1),
            'receive_emails' => $this->boolean()->defaultValue(1),
            'share_details' => $this->boolean()->defaultValue(1),

            'phone' => $this->string(50)->notNull(),
            'address1' => $this->string(100)->notNull(),
            'address2' => $this->string(100)->notNull(),
            'city' => $this->string(100)->notNull(),
            'state' => $this->string(100)->notNull(),
            'zip' => $this->string(20)->notNull(),
            'country' => $this->string(100)->notNull(),

            'created_at' => $this->integer()->notNull(),
            'updated_at' => $this->integer()->notNull(),
        ], $tableOptions);
        $this->addForeignKey('fk_user_profile_user_id', '{{%user_profile}}', 'user_id', '{{%user}}', 'id', 'CASCADE', 'CASCADE');

As you can see, it creates a foreign key mapping `user_profiles.user_id` to `user.id`. Nothing fancy there :)

However, Gii generated a `hasMany()` relation for it...

Since `user.id` is a primaryKey, Gii should recognize this and generate a `hasOne` relation, shouldn't it?

**Comment by samdark** [comment]
The structure reads as User hasMany profiles. Profile hasOne User. I think there's no way to detect that in your case there's only one profile.

**Comment by nineinchnick** [comment]
You have to have an unique key on the _foreign key columns_ for the relation to be hasOne. A primary key is also unique. And you can't create a foreign key pointing to a non-unique column at all.

**Comment by WadeShuler** [comment]
What if I added the `unique()` function to the integer columns that have relations?

ie:

    'user_id' => $this->integer()->notNull()->unique(),

Should it create a hasOne() relation then?

**Comment by nineinchnick** [comment]
@WadeShuler yes, that's what I've said.

**Pull request closed without merging**